## Практическое задание 3

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

#### Запишем сразу все функции для комбинаторики

Число **сочетаний** из $n$ элементов по $k$ элементов в каждом (в сочетаниях порядок не важен):

$$C_n^k = \frac{n!}{k!(n - k)!}$$

In [2]:
from math import factorial

In [3]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

Число **размещений** из $n$ элементов по $k$ элементов в каждом. При размещениях порядок важен, поэтому вариантов размещения может быть больше, чем сочетаний при заданных $k$  и $n$.

$$A_n^k = \frac{n!}{(n - k)!}$$

In [4]:
def arrangements(n, k):
    return int(factorial(n) / factorial(n - k))

Число **перестановок** из $n$ элементов — при перестановках важен порядок, но отличие от размещений в том, что применяются все имеющиеся $n$ элементов:

$$P_n = n!$$

In [5]:
def permutations(n):
    return int(factorial(n))

#### _1_ Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [6]:
salary = pd.DataFrame([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])

##### 1.1. Найдем средне арифметическое

In [7]:
mean_salary = salary.sum() / salary.count()
mean_salary

0    65.3
dtype: float64

Проверим

In [8]:
salary.mean()

0    65.3
dtype: float64

Видим, что средне арифметическое посчитали правильно

##### 1.2. Найдем среднее квадратичное отклонение

In [9]:
salary_std = np.sqrt(((salary - salary.mean())**2).sum() / (salary.count()-1))
salary_std

0    31.624607
dtype: float64

Проверим

In [10]:
salary.std(ddof=1)

0    31.624607
dtype: float64

Видим, что среднее квадратичное отклонение посчитали правильно

##### 1.3. Найдем смещенную оценку дисперсии

In [11]:
salary_variance = ((salary - salary.mean())**2).sum() / salary.count()
salary_variance

0    950.11
dtype: float64

Проверим

In [12]:
salary.var(ddof=0)

0    950.11
dtype: float64

Видим, что смещенную оценку дисперсии посчитали правильно

##### 1.4. Найдем несмещенную оценку дисперсии

In [13]:
salary_variance_2 = ((salary - salary.mean())**2).sum() / (salary.count()-1)
salary_variance_2

0    1000.115789
dtype: float64

Проверим

In [14]:
salary.var(ddof=1)

0    1000.115789
dtype: float64

Видим, что несмещенную оценку дисперсии посчитали правильно

#### _2_ В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. Какова вероятность того, что 3 мяча белые?

Это возможно в результате следующих исходов:  
1. Из первого ящика достали 1 белый мяч и 1 небелый, а из второго достали 2 белых и 2 небелых мяча.
2. Из первого ящика достали 2 белых мяча, а из второго достали 1 белый и 3 небелых мяча.
3. Из первого ящика достали 2 небелых мяча, а из второго достали 3 белых и 1 небелый мяч.  
Рассчитаем вероятность каждого исхода и сложим их.

Вероятность, того что мы из первого ящика достали 1 белый мяч и 1 небелый, а из второго достали 2 белых и 2 небелых мяча:

In [15]:
(combinations(5, 1)*combinations(3, 1)/combinations(8, 2))*(combinations(5, 2)*combinations(7, 2)/combinations(12, 4))

0.22727272727272727

Вероятность, того что мы из первого ящика достали 2 белых мяча, а из второго достали 1 белый и 3 небелых мяча:

In [16]:
combinations(5, 2)/combinations(8, 2)*(combinations(5, 1)*combinations(7, 3)/combinations(12, 4))

0.12626262626262627

Вероятность, того что мы из первого ящика достали 2 небелых мяча, а из второго достали 3 белых и 1 небелый мяч:

In [17]:
combinations(3, 2)/combinations(8, 2)*(combinations(5, 3)*combinations(7, 1)/combinations(12, 4))

0.01515151515151515

In [18]:
0.22727272727272727+0.12626262626262627+0.01515151515151515

0.3686868686868687

Таким образом, вероятность того, что 3 мяча белые = 0.3686868686868687

#### _3_ На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. Найти вероятность того, что выстрел произведен: a). первым спортсменом б). вторым спортсменом в). третьим спортсменом.

Воспользуемся для расчетов формулой Байеса $$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$
Событием $A$ будем считать попадание в мишень, а событиями $B_1, B_2$ и $B_3$ — что выстрел совершил первый, второй или третий спортсмен:


In [19]:
P_A_B_1 = 0.9
P_A_B_2 = 0.8
P_A_B_3 = 0.6
P_B_1 = 1/3
P_B_2 = 1/3
P_B_3 = 1/3

Вероятность попадания в мишень находим по формуле полной вероятности

In [20]:
P_A = 1/3*0.9+1/3*0.8+1/3*0.6
P_A

0.7666666666666666

3.1 Вероятность того, что стрелял первый спортсмен — при условии, что выстрел был удачным, — будем находить по формуле:

$$P(B_1\:|\:A) = \frac{P(B_1) \cdot P(A\:|\:B_1)}{P(A)}$$

---



In [21]:
P_B_1_A = P_B_1*P_A_B_1/P_A
P_B_1_A

0.391304347826087

3.2 Вероятность того, что стрелял второй спортсмен — при условии, что выстрел был удачным:

In [22]:
P_B_2_A = P_B_2*P_A_B_2/P_A
P_B_2_A

0.3478260869565218

3.3 Вероятность того, что стрелял третий спортсмен — при условии, что выстрел был удачным:

In [23]:
P_B_3_A = P_B_3*P_A_B_3/P_A
P_B_3_A

0.2608695652173913

#### _4_ В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: a). на факультете A б). на факультете B в). на факультете C?

Воспользуемся для расчетов формулой Бернулли $$P_n(X=k) = C_n^k p^k q^{n-k}$$


Пусть на A и B факультеты поступило по n студентов. Тогда на С факультет поступило 2n студентов 

4.1 Вероятность того, что студент учится на факультете A:  

In [24]:
P = combinations(1, 1)*(0.8**1)*(0.2**0)
P

0.8

Воспользуемся для расчетов формулой Байеса $$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$
Событием $A$ будем считать сдачу сессии, а событиями $B_1, B_2$ и $B_3$ — что сессию сдал студент факультета А, B или C:

In [25]:
P_A_B_1 = 0.8
P_A_B_2 = 0.7
P_A_B_3 = 0.9
P_B_1 = 0.25
P_B_2 = 0.25
P_B_3 = 0.5

Вероятность сдачи сессии находим по формуле полной вероятности

In [26]:
P_A = P_A_B_1*P_B_1+P_A_B_2*P_B_2+P_A_B_3*P_B_3
P_A

0.825

3.1 Вероятность того, что сессию сдал студннт факультета A будем находить по формуле:

$$P(B_1\:|\:A) = \frac{P(B_1) \cdot P(A\:|\:B_1)}{P(A)}$$

---



In [27]:
P_B_1_A = P_B_1*P_A_B_1/P_A
P_B_1_A

0.24242424242424246

3.2 Вероятность того, что сессию сдал студннт факультета B:

In [28]:
P_B_2_A = P_B_2*P_A_B_2/P_A
P_B_2_A

0.21212121212121213

3.3 Вероятность того, что сессию сдал студннт факультета C:

In [29]:
P_B_3_A = P_B_3*P_A_B_3/P_A
P_B_3_A

0.5454545454545455

#### _5_ Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя: а). все детали б). только две детали в). хотя бы одна деталь г). от одной до двух деталей?

In [30]:
p_1 = 0.1
p_2 = 0.2
p_3 = 0.25

##### 5.1 Вероятность того, что в первый месяц выйдут из строя все детали:

In [31]:
P_a = p_1*p_2*p_3
P_a

0.005000000000000001

##### 5.2 Вероятность того, что в первый месяц выйдут из строя только две детали:

Только две детали могут выйти из строя в следующих исходах:   
1. Из строя одновременно выйдет 1 и 2 детали, а третья будет работать
2. Из строя одновременно выйдет 1 и 3 детали, а вторая будет работать
3. Из строя одновременно выйдет 2 и 3 детали, а первая будет работать

1. Вероятность, что из строя одновременно выйдет 1 и 2 детали, а третья будет работать

In [32]:
P_b_1 = p_1*p_2*(1-p_3)
P_b_1

0.015000000000000003

2. Вероятность, что из строя одновременно выйдет 1 и 3 детали, а вторая будет работать

In [33]:
P_b_2 = p_1*p_3*(1-p_2)
P_b_2

0.020000000000000004

3. Вероятность, что из строя одновременно выйдет 2 и 3 детали, а первая будет работать

In [34]:
P_b_3 = p_2*p_3*(1-p_1)
P_b_3

0.045000000000000005

Вероятность того, что в первый месяц выйдут из строя только две детали равна суме вероятностей этих трех исходов

In [35]:
P_b = P_b_1 + P_b_2 + P_b_3
P_b

0.08000000000000002

##### 5.3 Вероятность того, что в первый месяц выйдет из строя хотя бы одна деталь:

Посчитаем от противного. Посчитаем вероятность, что из строя ни выйдет не одна деталь и вычтем эту вероятность из 1.

1. Вероятность, что из строя не выйдет 1 деталь:

In [36]:
P_c_1 = 1 - p_1
P_c_1

0.9

2. Вероятность, что из строя не выйдет 2 деталь:

In [37]:
P_c_2 = 1 - p_2
P_c_2

0.8

3. Вероятность, что из строя не выйдет 3 деталь:

In [38]:
P_c_3 = 1 - p_3
P_c_3

0.75

Вероятность, того что из строя не выйдет ни одна деталь равна произведению найденных вероятностей:

In [39]:
P_c_from_opposite = P_c_1 * P_c_2 * P_c_3
P_c_from_opposite

0.54

И, следовательно, вероятность, того что в первый месяц выйдет из строя хотя бы одна деталь равна:

In [40]:
P_c = 1 - P_c_from_opposite
P_c

0.45999999999999996

##### 5.4 Вероятность того, что в первый месяц выйдет из строя  от одной до двух деталей:

Данный вариант является результатом одного из исходов (логическое ИЛИ):
1. Из строя выйдет одна деталь
1. Из строя выйдут две детали

Вероятность того, что из строя выйдет одна деталь равна, а остальные будут работать:

In [41]:
P_d_1 = p_1*(1 - p_2)*(1 - p_3) + p_2*(1 - p_1)*(1 - p_3) + p_3*(1 - p_2)*(1 - p_1)
P_d_1

0.375

Вероятность того, что из строя выйдут две детали была посчитана выше и составила:

In [42]:
P_b

0.08000000000000002

Таким образом, вероятность того, что в первый месяц выйдет из строя от одной до двух деталей равна: 

In [43]:
P_d = P_d_1 + P_b
P_d

0.455